### PR TITLE
perf: speed up job bundle submission by reducing redundant stat calls

### PR DESCRIPTION
**Pull request by baxeaz**
### What was the problem/requirement? (What/Why)
When creating and uploading a job bundle performance is reduced by repeated stat calls to the same files

### What was the solution? (How)
Add a stat cache of previously retrieved stat results

### What is the impact of this change?
Faster job bundle creation.  In my own testing with Meerkat on top of the previously committed improvements I saw another ~20-28% improvement - reducing the creation time to 16-16.5 seconds.

### How was this change tested?
New unit tests added which pass
Integration tests pass
Manual testing

### Was this change documented?
Yes

### Does this PR introduce new dependencies?

This library is designed to be integrated into third-party applications that have bespoke and customized deployment environments. Adding dependencies will increase the chance of library version conflicts and incompatibilities. Please evaluate the addition of new dependencies. See the [Dependencies](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies) section of DEVELOPMENT.md for more details.

*   [ ] This PR adds one or more new dependency Python packages. I acknowledge I have reviewed the considerations for adding dependencies in [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies).
*   [ X] This PR does not add any new dependencies.

### Is this a breaking change?
No

### Does this change impact security?
No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/job_attachments/upload.py [hunk 1]
```diff
@@ -82,6 +83,50 @@
...
+
+    def get_stat(self, path: Path) -> Optional[os.stat_result]:
+        """Get cached stat result or perform stat and cache it"""
+        path_str = str(path)
+        if path_str not in self._cache:
+            try:
+                self._cache[path_str] = path.stat()
+            except (FileNotFoundError, PermissionError, OSError):
+                self._cache[path_str] = None
+        return self._cache[path_str]
```
Comment (jusiskin) [inline src/deadline/job_attachments/upload.py]: You could optimize this further using [`functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache).

Regardless of that, we should consider is having an unbounded cache - since this should be a best-effort optimization and not cause out of memory crashes.

There was a recent issue (#849) related to this same concern.

File: src/deadline/job_attachments/upload.py [hunk 2]
```diff
@@ -82,6 +83,50 @@
 
 logger = logging.getLogger("deadline.job_attachments.upload")
 
+
+class FileStatCache:
```
Comment (jusiskin) [inline src/deadline/job_attachments/upload.py]: This is becoming part of the [public API contract](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#public-contracts) based on the import path and class name. Is that intended?

**Review by jusiskin — CHANGES_REQUESTED** [review]

File: src/deadline/job_attachments/upload.py (hunk 1 above)
Comment (baxeaz) [reply src/deadline/job_attachments/upload.py]: Sure makes sense, switched and set a maxsize

File: src/deadline/job_attachments/upload.py (hunk 2 above)
Comment (baxeaz) [reply src/deadline/job_attachments/upload.py]: No reason for it to be currently - updated

File: src/deadline/job_attachments/upload.py [hunk 3]
```diff
@@ -1293,11 +1334,12 @@ def _find_matched_root_from_local_type_locations(
     def _get_total_size_of_files(self, paths: list[str]) -> int:
         def get_file_size(path_str: str) -> int:
             try:
-                return Path(path_str).resolve().stat().st_size
+                return self._stat_cache.get_size(Path(path_str))
```
Comment (jericht) [inline src/deadline/job_attachments/upload.py]: Double checked that `Path.stat()` follows symlinks by default, we are good to remove `.resolve()` call here: https://docs.python.org/3/library/pathlib.html#pathlib.Path.stat

File: src/deadline/job_attachments/upload.py [hunk 4]
```diff
@@ -82,6 +84,44 @@
...
+        # Fall back to direct exists() call if stat failed
+        return path.exists()
+
+    def is_dir(self, path: Path) -> bool:
+        """Check if path is directory, using cache when possible"""
+        stat_result = self._get_stat(str(path))
+        if stat_result is not None:
+            return stat.S_ISDIR(stat_result.st_mode)
+        # Fall back to direct is_dir() call if stat failed
+        return path.is_dir()
```
Comment (jericht) [inline src/deadline/job_attachments/upload.py:113]: Could we just return the stat result and the caller can query whatever information they want from it?

e.g. if the caller wants to know if it's a file, I think having another method for `is_file()` would be overcomplicating this class.

File: src/deadline/job_attachments/upload.py [hunk 5]
```diff
@@ -82,6 +84,44 @@
...
+class _FileStatCache:
+    """Private cache for file stat results to avoid redundant filesystem calls"""
+
+    @lru_cache(maxsize=1024)
+    def _get_stat(self, path_str: str) -> Optional[os.stat_result]:
+        """Get cached stat result for a path string"""
+        try:
+            return Path(path_str).stat()
+        except (FileNotFoundError, PermissionError, OSError):
+            return None
```
Comment (jericht) [inline src/deadline/job_attachments/upload.py:97]: Have we considered potential issues with stale data in the cache? (file metadata changes after the initial stat call) Let's double check that the lifetime of `S3AssetManager` instances is short enough that we aren't concerned about stale cache data.

Comment (baxeaz) [reply src/deadline/job_attachments/upload.py:97]: I have yes.  S3AssetManager lives for the duration of the submission.  Changing the data while submitting isn't supported, but wasn't supported previously either - we'd need to implement some other system to copy out or otherwise lock files from changes.

Comment (leongdl) [reply src/deadline/job_attachments/upload.py:97]: Just a thought, what if the cache was a nested element where the caller determines if it caches or not? Otherwise this decorator maybe forgotten and becomes a weird edge case as Jericho described.

File: test/unit/deadline/job_attachments/test_file_stat_cache.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+import stat
+from pathlib import Path
+from unittest.mock import patch, MagicMock
+
+from deadline.job_attachments.upload import _FileStatCache
+
+
+class TestFileStatCache:
```
Comment (leongdl) [inline test/unit/deadline/job_attachments/test_file_stat_cache.py:10]: One test related question - on windows we have some edge case code to handle long long paths, eg `\\c:\` when the system does not have the long path registry on. Would the cache class have any issues?

**Review by leongdl — COMMENTED** [review]
Nit comments only.

File: src/deadline/job_attachments/upload.py (hunk 4 above)
Comment (baxeaz) [reply src/deadline/job_attachments/upload.py:113]: We could do that - I've tried that change now for a bit, I'm not sure it actually comes out cleaner.  It ends up pushing some of these hidden internals out if you need to either handle the raised exceptions if I change it to truly just try to return stat or at least deal with the None case in the current form.  We also have unit tests that currently rely on this fallback behavior that would need to be changed to support that which isn't great.  Normally I'd agree we don't really want to have a wrapper for everything stat does, but because this is pretty clearly designed to be an internal helper for very specific use cases I'm a bit less concerned

File: src/deadline/job_attachments/upload.py (hunk 5 above)
Comment (baxeaz) [reply src/deadline/job_attachments/upload.py:97]: Not quite sure what you meant there - as in folks forget that it has the caching behavior?  You could always of course just call os.stat directly if you don't want to use the cache

File: test/unit/deadline/job_attachments/test_file_stat_cache.py (hunk 6 above)
Comment (baxeaz) [reply test/unit/deadline/job_attachments/test_file_stat_cache.py:10]: I could add a test if you can tell me a bit more about the case you're thinking of - at the end of the day though as long as we resolve the path in a uniform way it would return the string equivalent cache version - if it resolved to "something else" it would just be a cache miss and re resolve, no worse than it was before

File: src/deadline/job_attachments/upload.py (hunk 5 above)
Comment (leongdl) [reply src/deadline/job_attachments/upload.py:97]: Mostly from the caller side, not every caller will want cached data vs live data. So a boolean can easily select from both and versions as needed. 

Its ok , only thinking of code structure to reduce the number of if-else long term.

EG: 
`
if use_cache:
   do_this()
else
   do_that()
`

File: test/unit/deadline/job_attachments/test_file_stat_cache.py (hunk 6 above)
Comment (leongdl) [reply test/unit/deadline/job_attachments/test_file_stat_cache.py:10]: np - just thinking about all the cases we have to handle to not introduce regressions. The default case should have long path enabled so it is ok. Can be a long term follow up.

**Review by leongdl — DISMISSED** [review]

File: src/deadline/job_attachments/upload.py (hunk 5 above)
Comment (jusiskin) [reply src/deadline/job_attachments/upload.py:97]: It seems easy enough for a caller to not use the `_FileStatCache`. This seems like a convenience for the caller, so I'd not consider this a blocker.

File: src/deadline/job_attachments/upload.py [hunk 7]
```diff
@@ -82,6 +84,44 @@
...
+        return path.is_dir()
+
+    def get_size(self, path: Path) -> int:
+        """Get file size using cached stat"""
+        stat_result = self._get_stat(str(path))
+        if stat_result is not None:
+            return stat_result.st_size
+        # Log warning for missing files and return 0
+        logger.warning(f"Skipping file in size calculation: {path}")
+        return 0
```
Comment (jusiskin) [inline src/deadline/job_attachments/upload.py:122]: I'm curious about why we have fall-backs for `is_dir` and `exists` but not here for `get_size`?

Aren't all of these high-level wrappers for `stat()` calls? I'd imagine they would either all fail (regardless of calling `stat()` or the higher-level function) or succeed. The outcome for having fall-backs is that we pay a double `stat()` penalty when the underlying `stat()` call fails. Unless there is a timing issue where the first `stat()` fails, but the subsequent one may pass - but I'm wondering if this is something that was observed.

Maybe we should just remove the fallback logic and let the caller handle exceptions. What do you think?

Comment (baxeaz) [reply src/deadline/job_attachments/upload.py:122]: So if we remove the fallback logic at a minimum we'd need to rewrite one or two existing unit tests that rely on mocking exists calls - maybe that's ok, usually that would be something that might cause some pushback at review time though.  I could add the fallback logic to get_size as well though, that might be simpler - there's currently no use case that needs it but would keep things consistent?

Comment (baxeaz) [reply src/deadline/job_attachments/upload.py:122]: Talked offline - Removed the try/except from _get_total_size_of_files and added cache specific testing for this warning

**Review by leongdl — APPROVED** [review]

**Review by jusiskin — APPROVED** [review]
